umbraco/Umbraco-CMS
# Fix startup issues and timing with Examine and Nucache

This fixes #8893 and refactors #9859.

* Moves logic for cold boot into the components responsible for what needs to happen in cold boot. This used to be in callbacks in DatabaseServerMessenger.
* Removes the callbacks from DatabaseServerMessenger and obsoletes these callbacks.
* Ensure examine rebuilds don't occur until the first http request is done instead of on a timer which could be problematic with cold boots. A reason for the #8893 can be that during cold boots, the cold boot processing would take more than 5 seconds and then after those 5 seconds indexing occurs and this will happen before custom components execute that customize those indexes.
* PublishedSnapshotService now only initializes lazily when it's needed.
* Removes BatchedDatabaseServerMessenger.Startup, startup logic is just done in the component along with events.

The whole way that startup/init is done in v8 is not ideal but we are stuck with backwards compat.  Once main problems is that nucache (PublishedSnapshotService) does it's init in it's ctor which doesn't leave much flexibility with how the init process works. This will be fixed/changed in netcore (if not already, I know some of it is). 

## Testing

To force a cold boot, just delete the /TEMP/DistCache folder, then restart your app, this will cold boot.

* Ensure all indexes are rebuilt on cold boot
* Ensure indexes are not rebuilt on non-cold boot
* Ensure nucache is rebuilt (once) on a cold boot
* Ensure nucache is not rebuilt on non-cold boot
* Ensure that you can define custom Examine field types (i.e. #8893) on startup with a IUserComposer and those field types work both in a cold boot and non-cold boot
